sources/host-ctr/cmd/host-ctr/pid_namespaces.go

```diff
@@ -0,0 +1,184 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	"github.com/containerd/containerd"
+	"github.com/gofrs/flock"
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	// Get file descriptor that refers to a parent namespace, ioctl() request '_IO(NSIO, 0x2)'
+	nsGetParentNs        = 0xb702
+	lockfile      string = "/run/lock/host-ctr-pid-ns.lock"
+	manifestPath  string = "/etc/host-containers/info/pid-namespaces.json"
+)
+
+func getTaskPidNs(task *containerd.Task) (int64, error) {
+	pid := (*task).Pid()
+	pidNsFile, err := os.Open(fmt.Sprintf("/proc/%d/ns/pid", pid))
+	if err != nil {
+		return -1, fmt.Errorf("failed to open %d PID namespace file", pid)
+	}
+	pidNsFd := int(pidNsFile.Fd())
+
+	_, err = unix.IoctlGetInt(pidNsFd, nsGetParentNs)
+	if err == syscall.EPERM {
+		// If we can't access the parent PID namespace, this means the current container task is sharing the root PID
+		// namespace. We don't want to track the root PID namespace in the manifest, so we return -1.
+		return -1, nil
+	} else if err != nil {
+		return -1, errors.Wrap(err, fmt.Sprintf("ioctl failed to 'NS_GET_PARENTNS' for '%s'", pidNsFile.Name()))
+	}
+
+	// The PID namespace is represented by the inode number of the task's PID namespace file
+	var sb syscall.Stat_t
+	if err := syscall.Fstat(pidNsFd, &sb); err != nil {
+		return -1, errors.Wrap(err, "failed to stat task PID namespace fd")
+	}
+
+	return int64(sb.Ino), nil
+}
+
+// HostCtrPidNs represents the list of PID namespaces
+type HostCtrPidNs struct {
+	PidNamespaces []int64 `json:"pid-namespaces,omitempty"`
+}
+
+func addPidNsToManifest(pidNs int64) error {
+	fileLock := flock.New(lockfile)
```

Comment:
Why not locking the manifest file itself?

Reply:
Good point. I've always used a separate file for locking because of how we did it for the BR updates API flock. In that case we had to use a separate lock file because of other concerns I can remember at the moment.

Reply:
Actually, thinking about it a bit more. Is there any harm in maintaining a separate file for the lock? I don't see a compelling reason to lock on the file I'm trying to R/W from/to, I also have to create the path to the file and I'm not sure how ok it is for `flock` to be the one creating the file. In general, I feel safer keeping a different lock file, but that might be imprudent.

Reply:
See my comment below - `flock` is just a thin wrapper around a read / write mutex. I don't believe the file lock is  actually locking on files themselves. If all we care about is 1 static file, maybe we remove some of the cruft at just use and provision a read / write mutex ourselves.

Reply:
Edit: I don't think there's harm in managing a seperate lock file. And yes, good call on not letting flock attempt to create a file. I think that makes sense.

Reply:
Other than preventing concurrent writes from other processes while host-ctr is actually writing to the file, and not using yet another file in the `tmpfs` `/etc` filesystem, no. I just asked to understand if there was a deeper reason to create a separate file for the lock.

Reply:
> Other than preventing concurrent writes from other processes while host-ctr is actually writing to the file

Nothing prevents an outside process to write to the file even if it was `flock`ed by `host-ctr`.

From the man page:
>  flock() places advisory locks only; given suitable permissions on a file, a process is free to ignore the use of flock() and perform I/O on the file.

The `flock`ing is really only used for concurrency between `host-ctr` processes.

I elected to use a separate file for the filelock since I believe it's simpler (when thinking about which `host-ctr` is gonna flock first/create the file) and the only downside being needing to allocate an extra file descriptor/inode.